src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java

```diff
@@ -114,40 +136,142 @@ static SqlAuthenticationToken getSqlFedAuthTokenPrincipal(SqlFedAuthInfo fedAuth
-        } catch (MalformedURLException | InterruptedException e) {
+        } catch (InterruptedException e) {
             // re-interrupt thread
             Thread.currentThread().interrupt();
 
             throw new SQLServerException(e.getMessage(), e);
-        } catch (MsalThrottlingException | ExecutionException e) {
+        } catch (MalformedURLException | ExecutionException e) {
             throw getCorrectedException(e, aadPrincipalID, authenticationString);
         } finally {
             executorService.shutdown();
         }
     }
 
-    static SqlAuthenticationToken getSqlFedAuthTokenIntegrated(SqlFedAuthInfo fedAuthInfo,
+    static SqlAuthenticationToken getSqlFedAuthTokenPrincipalCertificate(SqlFedAuthInfo fedAuthInfo,
+            String aadPrincipalID, String certFile, String certPassword, String certKey, String certKeyPassword,
             String authenticationString) throws SQLServerException {
         ExecutorService executorService = Executors.newSingleThreadExecutor();
 
+        if (logger.isLoggable(Level.FINEST)) {
+            logger.finest(logContext + authenticationString + ": get FedAuth token for principal certificate: "
+                    + aadPrincipalID);
+        }
+
+        String defaultScopeSuffix = SLASH_DEFAULT;
+        String scope = fedAuthInfo.spn.endsWith(defaultScopeSuffix) ? fedAuthInfo.spn
+                                                                    : fedAuthInfo.spn + defaultScopeSuffix;
+        Set<String> scopes = new HashSet<>();
+        scopes.add(scope);
+
         try {
-            /*
-             * principal name does not matter, what matters is the realm name it gets the username in
-             * principal_name@realm_name format
-             */
-            KerberosPrincipal kerberosPrincipal = new KerberosPrincipal("username");
-            String user = kerberosPrincipal.getName();
+            ConfidentialClientApplication clientApplication = null;
+
+            // check if cert is PKCS12 first
+            try (InputStream is = new FileInputStream(certFile)) {
+                KeyStore keyStore = SQLServerCertificateUtils.loadPKCS12KeyStore(certFile, certPassword);
+
+                if (logger.isLoggable(Level.FINEST)) {
+                    logger.finest(logContext + "certificate type: " + keyStore.getType());
+
+                    // we don't need to do this unless logging enabled since MSAL will fail if cert is not valid
+                    Enumeration<String> enumeration = keyStore.aliases();
+                    while (enumeration.hasMoreElements()) {
+                        String alias = (String) enumeration.nextElement();
+                        X509Certificate cert = (X509Certificate) keyStore.getCertificate(alias);
+                        cert.checkValidity();
+                        logger.finest(logContext + "certificate: " + cert.toString());
```

Comment:
Unnecessary toString().
```suggestion
                        logger.finest(logContext + "certificate: " + cert);
```